Core/Core/Courses/CourseListView.swift

```diff
@@ -232,3 +275,49 @@ struct CourseListView_Previews: PreviewProvider {
     }
 }
 #endif
+
+@available(iOSApplicationExtension 13.0, *)
+struct TagPrefKey: PreferenceKey, Equatable {
+    struct Tag: Equatable, CustomStringConvertible {
+        var name: String
+        var subtags: [Tag]
+
+        var description: String { description().joined(separator: "\n") }
+        private func description(_ indent: String = "") -> [String] {
+            ["\(indent)\(name)"] + subtags.flatMap {
+                $0.description("\(indent)  ")
+            }
+        }
+    }
+
+    typealias Value = [Tag]
+    static let defaultValue: Value = []
+    static func reduce(value: inout Value, nextValue: () -> Value) {
+        value += nextValue()
+    }
+}
+
+@available(iOSApplicationExtension 13.0, *)
+struct TestID: ViewModifier {
+    // This really doesn't have to be string, and could be a more useful type
+    let name: String
+
+    func body(content: Content) -> some View {
+        content.transformPreference(TagPrefKey.self) { tags in
+            tags = [TagPrefKey.Tag(name: self.name, subtags: tags)]
+        }
+    }
+}
+
+@available(iOSApplicationExtension 13.0, *)
+extension View {
+    #if DEBUG
+    func testID(_ name: @escaping @autoclosure () -> Any? = nil) -> some View {
+        self.modifier(TestID(name: "\(name() ?? type(of: self))"))
```

Comment:
so in debug mode, the testID modifier updates a preference to keep track of the heirarchy? That's pretty clever. In prod this is an inlinable no-op. Theoretically we can put whatever info we want in the tags too.